java-squid/src/main/java/org/sonar/java/se/ProgramState.java

```diff
@@ -85,16 +82,21 @@ static ProgramState put(ProgramState programState, Symbol symbol, SymbolicValue
     if (symbol.isUnknown()) {
       return programState;
     }
-    SymbolicValue symbolicValue = programState.values.get(symbol);
-    // update program state only for a different symbolic value
-    if (symbolicValue == null || !symbolicValue.equals(value)) {
-      Map<Symbol, SymbolicValue> temp = Maps.newHashMap(programState.values);
-      temp.put(symbol, value);
-      return new ProgramState(temp, programState.constraints, programState.visitedPoints, programState.stack);
+    PMap<Symbol, SymbolicValue> newValue = programState.values.put(symbol, value);
+    if(newValue != programState.values) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)